packages/react-core/src/components/Menu/MenuItem.tsx

```diff
@@ -77,20 +80,80 @@ const MenuItemBase: React.FunctionComponent<MenuItemProps> = ({
   icon,
   actions,
   onShowFlyout,
-  innerRef,
   drilldownMenu,
   isOnPath,
   ...props
 }: MenuItemProps) => {
+  const {
+    menuId,
+    parentMenu,
+    onSelect,
+    onActionClick,
+    activeItemId,
+    selected,
+    drilldownItemPath,
+    onDrillIn,
+    onDrillOut
+  } = React.useContext(MenuContext);
   const Component = component || to ? 'a' : 'button';
   const [flyoutVisible, setFlyoutVisible] = React.useState(false);
   const [flyoutTarget, setFlyoutTarget] = React.useState(null);
+  const flyoutContext = React.useContext(FlyoutContext);
+  const [flyoutXDirection, setFlyoutXDirection] = React.useState(flyoutContext.direction);
+  const [offsetX, setOffsetX] = React.useState(0);
+  const [offsetY, setOffsetY] = React.useState(0);
+  const ref = React.useRef<HTMLLIElement>();
 
+  const hasFlyout = flyoutMenu !== undefined;
   const showFlyout = (displayFlyout: boolean) => {
     setFlyoutVisible(displayFlyout);
     onShowFlyout && displayFlyout && onShowFlyout();
   };
 
+  React.useLayoutEffect(() => {
+    if (hasFlyout && ref.current && canUseDOM) {
+      const flyoutMenu = ref.current.lastElementChild;
+      if (flyoutMenu && flyoutMenu.classList.contains(styles.menu)) {
+        const origin = ref.current.getClientRects()[0];
+        const rect = flyoutMenu.getClientRects()[0];
+        if (origin && rect) {
+          const spaceLeftLeft = origin.x - rect.width;
+          const spaceLeftRight = window.innerWidth - origin.x - origin.width - rect.width;
+          if (spaceLeftLeft < 0 && spaceLeftRight < 0) {
+            // Calculate x offsets
+            setOffsetX(-spaceLeftLeft);
+          } else if (spaceLeftRight < 0) {
+            setFlyoutXDirection('left');
+          } else if (spaceLeftLeft < 0) {
+            setFlyoutXDirection('right');
+          }
+          const spaceLeftBot = window.innerHeight - origin.y - rect.height;
+          // This is calculated with respect to spaceLeftBot
+          const spaceLeftTop = window.innerHeight - rect.height;
+          if (spaceLeftTop < 0 && spaceLeftBot < 0) {
+            // Idea: can scroll down, but not up
+            setOffsetY(0);
+          } else if (spaceLeftBot < 0) {
+            setOffsetY(-spaceLeftBot);
+          }
+        }
+      }
+    }
+  }, [flyoutVisible, flyoutMenu]);
+
+  React.useEffect(() => {
+    const flyoutMenu = ref.current.lastElementChild as HTMLElement;
+    if (flyoutMenu && flyoutMenu.classList.contains(styles.menu)) {
+      flyoutMenu.style.right = flyoutXDirection === 'right' ? 'auto' : `calc(100% - ${offsetX}px`;
+      flyoutMenu.style.left = flyoutXDirection === 'left' ? 'auto' : `calc(100% - ${offsetX}px`;
```

Comment:
see if you can use `--pf-c-menu--m-flyout__menu--left-offset` here

Reply:
Unfortunately left/right offsets don't work well since I'm not using the `pf-m-menu-left` class. I can't use the class since if I do apply it _to the parent_ the consumer might just override it with some bit of state inside the menu.

Reply:
Ah! Sounds like I could just create generic left/right offset vars that aren't attached to a class?

Though `--pf-c-menu--m-flyout__menu--left-offset` isn't attached to a class, so I would expect that to work. `--pf-c-menu--m-flyout__list-item--m-menu-left__menu--right-offset` works via `.pf-m-menu-left`, but I can update that to just be `--pf-c-menu--m-flyout__menu--right-offset`

Reply:
I think I could use the CSS vars and add the class if this selector was just `.pf-c-menu.pf-c-menu-left`: https://github.com/patternfly/patternfly/blob/main/src/patternfly/components/Menu/menu.scss#L184